File: ios/AirshipFrameworkProxy/TagGroupOperation.swift
/* Copyright Airship and Contributors */

import Foundation

#if canImport(AirshipKit)
import AirshipKit
#elseif canImport(AirshipCore)
import AirshipCore
#endif

public struct TagGroupOperation: Decodable {
    enum Action: String, Codable {
        case setTags = "set"
        case removeTags = "remove"
        case addTags = "add"
    }

    let action: Action
    let tags: [String]
    let group: String

    private enum CodingKeys: String, CodingKey {
        case group = "group"
        case action = "operationType"
        case tags = "tags"
    }

    func apply(editor: TagGroupsEditor) {
        switch(action) {
        case .removeTags:
            editor.remove(tags, group: group)
        case .setTags:
            editor.set(tags, group: group)
        case .addTags:
            editor.add(tags, group: group)
        }
    }
}

File: android/airship-framework-proxy/src/main/java/com/urbanairship/android/framework/proxy/TagGroupOperation.kt
package com.urbanairship.android.framework.proxy

import com.urbanairship.channel.TagGroupsEditor
import com.urbanairship.json.JsonMap

public enum class TagGroupOperationAction {
    ADD, REMOVE, SET
}

public data class TagGroupOperation(
    public val group: String,
    public val tags: List<String>,
    public val action: TagGroupOperationAction
) {
    public constructor(json: JsonMap) : this(
        group = json.require("group").requireString(),
        tags = json.require("tags").requireList().map { it.requireString() },
        action = TagGroupOperationAction.valueOf(
            json.require("operationType").requireString().uppercase()
        )
    )
}

internal fun TagGroupOperation.applyOperation(editor: TagGroupsEditor) {
    when (this.action) {
        TagGroupOperationAction.ADD -> editor.addTags(this.group, this.tags.toSet())
        TagGroupOperationAction.REMOVE -> editor.removeTags(this.group, this.tags.toSet())
        TagGroupOperationAction.SET -> editor.setTags(this.group, this.tags.toSet())
    }
}
